topostats/tracing/disordered_tracing.py

```diff
@@ -254,6 +255,31 @@ def smooth_mask(
         LOGGER.debug(f"[{self.filename}] : smoothing done by dilation {dilation_iterations}")
         return self.re_add_holes(grain, dilation, holearea_min_max)
 
+    @staticmethod
+    def calculate_dna_width(smoothed_mask: npt.NDArray, pruned_skeleton: npt.NDArray, px2nm: float = 1) -> float:
+        """
+        Calculate the mean width in metres of the DNA using the trace and mask.
+
+        Parameters
+        ----------
+        smoothed_mask : npt.NDArray
+            Smoothed mask to be measured.
+        pruned_skeleton : npt.NDArray
+            Pruned skeleton.
+        px2nm : float
+            Scaling of pixels to nanometres.
+
+        Returns
+        -------
+        float
+            Width of grain in metres.
+        """
+        # Code will go here
+        dist_trans = distance_transform_edt(smoothed_mask)
+        comb = np.where(pruned_skeleton == 1, dist_trans, 0)
+
+        return comb[comb != 0].mean() * 2 * px2nm
```

Comment:
In my confusion I've now looked up what [`distance_transform_edt()`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.distance_transform_edt.html) does and am curious why the mean value is multiplied by `* 2` here?

A dummy example which hopefully explains my concern.

```
import numpy as np
from skimage.morphology import skeletonize
from scipy import ndimage

mask = np.asarray([
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 1, 1, 1, 1, 1, 1, 1, 1, 0],
    [0, 1, 1, 1, 1, 1, 1, 1, 1, 0],
    [0, 1, 1, 0, 0, 0, 1, 1, 1, 0],
    [0, 1, 1, 0, 0, 0, 1, 1, 1, 0],
    [0, 1, 1, 0, 0, 0, 1, 1, 1, 0],
    [0, 1, 1, 1, 1, 1, 1, 1, 1, 0],
    [0, 1, 1, 1, 1, 1, 1, 1, 1, 0],
    [0, 1, 1, 1, 1, 1, 1, 1, 1, 0],
    [0, 1, 1, 1, 1, 1, 1, 1, 1, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],])
skeleton = np.asarray(skeletonize(mask), dtype=int)
distance = ndimage.distance_transform_edt(mask)
comb = np.where(skeleton ==1, distance, 0)
```

The resulting arrays are...

## `skeleton`

```
array([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 1, 1, 1, 1, 1, 0, 0, 0],
       [0, 1, 0, 0, 0, 0, 0, 1, 0, 0],
       [0, 1, 0, 0, 0, 0, 0, 1, 0, 0],
       [0, 1, 0, 0, 0, 0, 0, 1, 0, 0],
       [0, 1, 0, 0, 0, 0, 0, 1, 0, 0],
       [0, 1, 0, 0, 0, 0, 0, 1, 0, 0],
       [0, 0, 1, 1, 1, 1, 1, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]])
```

## `distance`

```
array([[0. ,0. , 0.  , 0., 0.,  0., 0.  , 0., 0., 0.],
       [0. ,1. , 1.  , 1., 1.,  1., 1.  , 1., 1., 0.],
       [0. ,1. , 1.41, 1., 1.,  1., 1.41, 2., 1., 0.],
       [0. ,1. , 1.  , 0., 0.,  0., 1.  , 2., 1., 0.],
       [0. ,1. , 1.  , 0., 0.,  0., 1.  , 2., 1., 0.],
       [0. ,1. , 1.  , 0., 0.,  0., 1.  , 2., 1., 0.],
       [0. ,1. , 1.41, 1., 1.,  1., 1.41, 2., 1., 0.],
       [0. ,1. , 2.  , 2., 2.,  2., 2.23, 2., 1., 0.],
       [0. ,1. , 2.  , 2., 2.,  2., 2.  , 2., 1., 0.],
       [0. ,1. , 1.  , 1., 1.,  1., 1.  , 1., 1., 0.],
       [0. ,0. , 0.  , 0., 0.,  0., 0.  , 0., 0., 0.]])
```

## `comb`

```
array([[0., 0., 0., 0., 0., 0., 0.  , 0., 0., 0.],
       [0., 0., 1., 1., 1., 1., 1.  , 0., 0., 0.],
       [0., 1., 0., 0., 0., 0., 0.  , 2., 0., 0.],
       [0., 1., 0., 0., 0., 0., 0.  , 2., 0., 0.],
       [0., 1., 0., 0., 0., 0., 0.  , 2., 0., 0.],
       [0., 1., 0., 0., 0., 0., 0.  , 2., 0., 0.],
       [0., 1., 0., 0., 0., 0., 0.  , 2., 0., 0.],
       [0., 0., 2., 2., 2., 2., 2.23, 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.  , 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.  , 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.  , 0., 0., 0.]])

```


Looking at the right hand-side of this square the `comb` has a width of `2` where the mask width is `3` pixels. By multiplying by `2` it implies the width is `4` when actually it is `3`. This in turn inflates the mean width.

Reply:
I think this would apply to all odd pixel width objects, that we would overcompensate by one 1 pixel? This should only be a big problem for very low resolution scans with the error essentially just being the 1*px_nm_scaling. Having discussed with the group we have decided this is acceptable for now. When @llwiggins and @SylviaWhittle write the grainstats documentation I will let them know so that they can add a note to say that cation should be taken with this parameter for low resolution scans. There are plans for a more sophisticated function to calculate in the future, but this is lower priority.

Reply:
Ok, I'm more than happy to have incremental improvements and caveats with existing methods noted, nothing is ever perfect and we move forward in small steps, not leaps and bounds.

Prior to merging I will...

- add a modified version of the above test with the true width being `3` (or another odd number) and mark it with `@pytest.mark.xfail()` which means that we will expect it to fail on masks with an odd-numbered of pixel widths.
- write an issue to address this in the future so it doesn't get forgotten about.